executor/src/witgen/jit/witgen_inference.rs

```diff
@@ -303,40 +311,61 @@ impl<'a, T: FieldElement, FixedEval: FixedEvaluator<T>> WitgenInference<'a, T, F
         }
     }
 
-    fn ingest_effects(&mut self, process_result: ProcessResult<T, Variable>) -> ProcessSummary {
+    /// Analyze the effects and update the internal state.
+    /// If the effect is the result of a machine call, `identity_id` must be given
+    /// to avoid two calls to the same sub-machine on the same row.
+    /// Returns true if there was progress.
+    fn ingest_effects(
+        &mut self,
+        process_result: ProcessResult<T, Variable>,
+        identity_id: Option<(u64, i32)>,
+    ) -> bool {
         let mut progress = false;
         for e in process_result.effects {
             match &e {
                 Effect::Assignment(variable, assignment) => {
-                    assert!(self.known_variables.insert(variable.clone()));
```

Comment:
This is so that so that what we can keep assigning to a variable, refining the range constraints, right?

Reply:
Maybe we can keep the old version, this should actually not happen.

Reply:
I'm wrong. It does not happen in this PR but in the next: If we process an identity in the way where we only consider concrete variables as known, we can get an assignment to an already known (but not concrete) variable.